
-----

## 📊 Cálculo de Taxa de Conversão e Impacto de Mudanças

> Uma empresa de tecnologia quer avaliar o impacto de uma nova interface em seu aplicativo. Foi realizado um experimento onde um grupo de usuários utilizou a versão antiga (grupo controle) e outro grupo usou a nova versão (grupo de teste).
>
> Os resultados coletados foram os seguintes:

<br>

<center>

| Grupo | Número de Usuários | Número de Ações Completadas | Taxa de Conversão (%) |
| :---: | :---: | :---: | :---: |
| **Controle** | 5.000 | 1.200 | **?** |
| **Teste** | 5.000 | 1.500 | **?** |

</center>

<br>

-----

### 🎯 Desafios:

  * **a)** Calcule a taxa de conversão para cada grupo.
  * **b)** Determine o percentual de aumento ou redução na conversão com a nova interface.
  * **c)** Se o critério para adoção da nova interface for um aumento mínimo de **20% na taxa de conversão**, a mudança deve ser implementada?

In [1]:
numero_usuario_control = 5000
numero_usuario_teste = 5000
numero_acao_comple_cont = 1200
numero_acao_comple_test = 1500

import pandas as pd

**a)** Calcule a taxa de conversão para cada grupo.

In [2]:
taxa_conv_controle = (numero_acao_comple_cont/numero_usuario_control)
taxa_conv_teste = (numero_acao_comple_test/numero_usuario_teste)

print(f"Taxa de conversão Controle: {taxa_conv_controle:.0%}")
print(f"Taxa de conversão Teste: {taxa_conv_teste:.0%}")

Taxa de conversão Controle: 24%
Taxa de conversão Teste: 30%


**b)** Determine o percentual de aumento ou redução na conversão com a nova interface.

In [3]:
percentual_mudanca = (taxa_conv_teste-taxa_conv_controle)/taxa_conv_controle
print(f"Taxa de converção: {percentual_mudanca:.0%}")

Taxa de converção: 25%


**c)** Se o critério para adoção da nova interface for um aumento mínimo de **20% na taxa de conversão**, a mudança deve ser implementada?

In [4]:
if percentual_mudanca >= 0.20:
    print("Implementar mudança de interface")
else:
    print("Não implementar mudança de interface")

Implementar mudança de interface


---
## 📈 Exercício 2: Teste A/B e Análise de Resultados

> Uma empresa de comércio eletrônico está realizando um teste A/B para avaliar qual versão de um anúncio gera mais cliques. A empresa coletou os seguintes dados:

<br>

<center>

| Versão do Anúncio | Impressões | Cliques | Taxa de Cliques (%) |
| :---: | :---: | :---: | :---: |
| **Anúncio A** | 50.000 | 2.000 | **?** |
| **Anúncio B** | 50.000 | 2.300 | **?** |

</center>

<br>

### 🎯 Desafios:

* **a)** Calcule a taxa de cliques para cada versão do anúncio.
* **b)** Qual versão teve melhor desempenho?
* **c)** Se o nível de significância adotado para o experimento for **5% (α=0.05)** e o **p-valor** calculado for **0,08**, podemos afirmar que a diferença entre os anúncios é estatisticamente significativa?
---

In [5]:
impressoes_A = 50000
impressoes_B = 50000
cliques_A = 2000
cliques_B = 2300

**a)** Calcule a taxa de cliques para cada versão do anúncio.

In [6]:
taxa_de_cliques_A = cliques_A/impressoes_A 
taxa_de_cliques_B = cliques_B/impressoes_B

print(f"Taxa de cliques A: {taxa_de_cliques_A:.1%}")
print(f"Taxa de cliques B: {taxa_de_cliques_B:.1%}")

Taxa de cliques A: 4.0%
Taxa de cliques B: 4.6%


**b)** Qual versão teve melhor desempenho?

In [7]:
print(f"b) Avaliação de Desempenho:")
if taxa_de_cliques_B > taxa_de_cliques_A:
    print(f"   - O Anúncio B teve um desempenho melhor ({taxa_de_cliques_B:.2%}) do que o Anúncio A ({taxa_de_cliques_A:.2%}).")
elif taxa_de_cliques_A > taxa_de_cliques_B:
    print(f"   - O Anúncio A teve um desempenho melhor ({taxa_de_cliques_A:.2%}) do que o Anúncio B ({taxa_de_cliques_B:.2%}).")
else:
    print("   - Ambos os anúncios tiveram o mesmo desempenho.")


b) Avaliação de Desempenho:
   - O Anúncio B teve um desempenho melhor (4.60%) do que o Anúncio A (4.00%).


**c)** Se o nível de significância adotado para o experimento for **5% (α=0.05)** e o **p-valor** calculado for **0,08**, podemos afirmar que a diferença entre os anúncios é estatisticamente significativa?

In [8]:
p_valor_do_teste = 0.08
nivel_de_significancia = 0.05

print(f"   - P-Valor do teste: {p_valor_do_teste}")
print(f"   - Nível de Significância (α): {nivel_de_significancia}")

# A "Regra de Ouro" da estatística:
# Se o p-valor for MENOR ou IGUAL ao nível de significância, o resultado é significativo.
if p_valor_do_teste <= nivel_de_significancia:
    print("\n   -> Conclusão: SIM, a diferença é estatisticamente significativa.")
    print("      (Podemos afirmar com confiança que o Anúncio B é realmente melhor.)")
else:
    print("\n   -> Conclusão: NÃO, a diferença NÃO é estatisticamente significativa.")
    print("      (Apesar do Anúncio B parecer melhor, não temos evidências fortes o suficiente.")
    print("      (A diferença pode ter ocorrido apenas por sorte ou acaso.)")

   - P-Valor do teste: 0.08
   - Nível de Significância (α): 0.05

   -> Conclusão: NÃO, a diferença NÃO é estatisticamente significativa.
      (Apesar do Anúncio B parecer melhor, não temos evidências fortes o suficiente.
      (A diferença pode ter ocorrido apenas por sorte ou acaso.)


---
## 🤖 Exercício 3: Interpretação de Teste Estatístico

> Uma empresa realizou um teste para verificar se a inclusão de um chatbot em seu site aumentaria a taxa de fechamento de vendas. Os resultados foram analisados utilizando um teste t de Student, e os seguintes valores foram obtidos:
>
> * **Média de vendas no grupo controle:** 5,2
> * **Média de vendas no grupo teste:** 6,1
> * **Desvio padrão:** 1,4
> * **Número de participantes em cada grupo:** 1.000
> * **p-valor:** 0,03

<br>

### 🎯 Desafios:

* **a)** O aumento na média de vendas foi estatisticamente significativo considerando um nível de significância de **5%**?
* **b)** Explique o significado do p-valor e como ele deve ser interpretado.
---

In [9]:
# p-valor: O resultado que o teste estatístico nos deu.
p_valor = 0.03

# nível_significancia (Alfa): O nosso "limite de dúvida" aceitável.
# O problema definiu como 5% (ou 0.05).
nivel_significancia = 0.05

print(f"--- Análise do Teste do Chatbot ---")
print(f"P-Valor obtido: {p_valor}")
print(f"Nível de Significância (α) definido: {nivel_significancia}")
print("-" * 100 + "\n")

# Esta é a "Regra de Ouro" da estatística:
# Se o P-Valor for MENOR ou IGUAL ao Nível de Significância,
# nós rejeitamos a "sorte" e aceitamos que o resultado é REAL.

if p_valor <= nivel_significancia:
    
    print("a) Conclusão: SIM, o aumento na média de vendas é estatisticamente significativo.")
    print("   (A evidência sugere que o Chatbot realmente causou o aumento nas vendas.)")
else:
   
    print("a) Conclusão: NÃO, o aumento na média de vendas NÃO é estatisticamente significativo.")
    print("   (Não podemos afirmar que o Chatbot causou o aumento; a diferença pode ter sido apenas sorte.)")



--- Análise do Teste do Chatbot ---
P-Valor obtido: 0.03
Nível de Significância (α) definido: 0.05
----------------------------------------------------------------------------------------------------

a) Conclusão: SIM, o aumento na média de vendas é estatisticamente significativo.
   (A evidência sugere que o Chatbot realmente causou o aumento nas vendas.)


**b)** Explique o significado do p-valor e como ele deve ser interpretado.

In [10]:
print(f"""
   O p-valor ({p_valor}, ou 3%) representa a probabilidade de obter um
   resultado tão extremo quanto o observado (um aumento de 5.2 para 6.1),
   partindo do princípio de que o chatbot NÃO tivesse efeito nenhum
   (ou seja, se a 'hipótese nula' fosse verdadeira).
""")
print("   Em termos simples: É a 'probabilidade de ter sido apenas sorte'.")
print("   Como a chance de ser sorte (3%) é muito baixa (menor que nosso limite de 5%),")
print("   nós concluímos que NÃO foi sorte.")


   O p-valor (0.03, ou 3%) representa a probabilidade de obter um
   resultado tão extremo quanto o observado (um aumento de 5.2 para 6.1),
   partindo do princípio de que o chatbot NÃO tivesse efeito nenhum
   (ou seja, se a 'hipótese nula' fosse verdadeira).

   Em termos simples: É a 'probabilidade de ter sido apenas sorte'.
   Como a chance de ser sorte (3%) é muito baixa (menor que nosso limite de 5%),
   nós concluímos que NÃO foi sorte.


---
## 🔬 Exercício 4: Planejamento de um Experimento e Cálculo de Amostra

> Uma plataforma de cursos online deseja testar se uma nova recomendação de vídeos aumenta o tempo médio assistido pelos usuários. Atualmente, os alunos assistem, em média, **35 minutos** de aula por dia, com um desvio padrão de **10 minutos**. O time de análise deseja conduzir um experimento com um nível de confiança de **95%** e uma margem de erro de **2 minutos**.
>
> Utilize a fórmula abaixo para calcular o tamanho da amostra:
>
> $$n=\frac{Z^{2}\cdot\sigma^{2}}{e^{2}}$$
>
> **Onde:**
> * $Z=1,96$ (para 95% de confiança)
> * $\sigma=10$ (desvio padrão)
> * $e=2$ (margem de erro)

<br>

### 🎯 Desafios:

* **a)** Qual deve ser o tamanho mínimo da amostra para garantir resultados confiáveis?
* **b)** Se a empresa quiser reduzir a margem de erro para **1,5 minutos**, como isso afetaria o tamanho da amostra?
---

**a)** Qual deve ser o tamanho mínimo da amostra para garantir resultados confiáveis?

In [11]:
import math

z = 1.96
desvio_padrao = 10
margem_de_erro = 2

n = (z**2 * desvio_padrao**2) / margem_de_erro**2

print(f"O valor exato de n é: {n}")
# O tamanho da amostra deve ser um número inteiro,
# então arredondamos para cima.
n_arredondado = math.ceil(n)
print(f"O tamanho mínimo da amostra é: {n_arredondado}")

O valor exato de n é: 96.03999999999999
O tamanho mínimo da amostra é: 97


**b)** Se a empresa quiser reduzir a margem de erro para **1,5 minutos**, como isso afetaria o tamanho da amostra?

In [12]:
margem_de_erro = 1.5

n2 = (z**2 * desvio_padrao**2) / margem_de_erro**2

print(f"O valor exato de n é: {n2}")
# O tamanho da amostra deve ser um número inteiro,
# então arredondamos para cima.
n_arredondado2 = math.ceil(n2)
print(f"O tamanho mínimo da amostra para margem de erro em 1,5 é: {n_arredondado2}")
print(f"Se a empresa busca diminuir a margem de erro, ela devera aumentar a amostra em: {(n_arredondado2 - n_arredondado) / n_arredondado:.2%}")

O valor exato de n é: 170.73777777777775
O tamanho mínimo da amostra para margem de erro em 1,5 é: 171
Se a empresa busca diminuir a margem de erro, ela devera aumentar a amostra em: 76.29%


---
## 📧 Exercício 5: Estudo de Caso - Melhorando a Retenção de Usuários

> Uma empresa de streaming percebeu que muitos usuários cancelam suas assinaturas após três meses de uso. Para tentar melhorar a retenção, a empresa testou uma nova estratégia de envio de e-mails personalizados. O experimento teve os seguintes resultados:

<br>

<center>

| Grupo | Número de Usuários | Usuários que Continuaram Assinatura | Taxa de Retenção (%) |
| :---: | :---: | :---: | :---: |
| **Controle** | 10.000 | 3.200 | **?** |
| **Teste** | 10.000 | 3.600 | **?** |

</center>

<br>

### 🎯 Desafios:

* **a)** Calcule a taxa de retenção para cada grupo.
* **b)** A nova estratégia de e-mails personalizados teve um impacto positivo?
* **c)** Que outros fatores devem ser analisados antes de implementar essa estratégia para todos os usuários?
---

In [13]:
controle_usuario = 10000
teste_usuario = 10000
controle_assinaturas = 3200
teste_assinaturas = 3600

**a)** Calcule a taxa de retenção para cada grupo.

In [14]:
rentencao_controle = controle_assinaturas / controle_usuario
rentencao_teste = teste_assinaturas / teste_usuario

print(f"Taxa de remtenção do Grupo de Controle: {rentencao_controle:.1%}")
print(f"Taxa de remtenção do Grupo de Teste: {rentencao_teste:.1%}")

Taxa de remtenção do Grupo de Controle: 32.0%
Taxa de remtenção do Grupo de Teste: 36.0%


**b)** A nova estratégia de e-mails personalizados teve um impacto positivo?

Sim

**c)** Que outros fatores devem ser analisados antes de implementar essa estratégia para todos os usuários?


---

### 1. 🔬 Significância Estatística

Este é o fator mais importante. A diferença de 4% foi real ou apenas **sorte (acaso)**?

* **O que analisar:** Você precisa calcular o **p-valor** do teste.
* **Por quê:** O grupo de teste tinha 10.000 usuários[cite: 178]. É um bom número, mas pode ter acontecido de, *por puro acaso*, os usuários desse grupo serem um pouco mais engajados. Se o p-valor for alto (geralmente maior que 0,05), a diferença não é confiável, e o resultado pode ter sido apenas sorte.

---

### 2. 💸 Análise de Custo vs. Benefício (ROI)

A nova estratégia deu mais lucro do que custou?

* **O que analisar:**
    * **Custo:** Quanto custa implementar e manter essa nova estratégia? (Ex: custo de software de personalização, horas de engenharia, tempo da equipe de marketing).
    * **Benefício:** Qual o valor monetário dos **400 usuários extras** retidos (3.600 - 3.200)[cite: 178]? Quanto eles pagam de assinatura?
* **Por quê:** Se a nova estratégia de e-mail custa R$ 50.000 por mês para rodar, mas o lucro dos 400 usuários retidos for de apenas R$ 30.000, a empresa estará **perdendo dinheiro**, mesmo que a retenção tenha aumentado.

---

### 3. 📉 Métricas Negativas (ou Métricas "Guardiãs")

A nova estratégia melhorou a retenção, mas **piorou alguma outra coisa**?

* **O que analisar:**
    * **Taxa de Abertura:** Os usuários abriram menos os e-mails?
    * **Taxa de Cliques:** Eles pararam de clicar?
    * **Taxa de Descadastro (Unsubscribe):** Esta é a mais perigosa. Se a nova estratégia fez muito mais gente se descadastrar da lista de e-mails, você pode ter melhorado a retenção de 3 meses, mas perdeu a capacidade de se comunicar com esses usuários para sempre.
* **Por quê:** Você não pode "vencer a batalha" (reter 4% a mais) e "perder a guerra" (fazer 10% dos seus usuários bloquearem seus e-mails).

---

### 4. 👥 Análise de Segmentos

A melhoria foi igual para todos os tipos de usuários?

* **O que analisar:** Quebre os resultados (32% vs 36%) por diferentes grupos:
    * Usuários novos vs. Usuários antigos
    * Usuários por país ou região
    * Usuários que assistem gêneros diferentes (Ex: comédia vs. drama)
* **Por quê:** Você pode descobrir que a estratégia foi **excelente** para usuários novos (aumentou 10% a retenção deles), mas foi **péssima** para usuários antigos (diminuiu 2% a retenção deles). Se for esse o caso, você não deve implementar para todos, apenas para o segmento que se beneficiou.

---

### 5. ⚙️ Viabilidade Técnica e Escalabilidade

O que funcionou para 10.000 usuários vai funcionar para 10 milhões?

* **O que analisar:** O sistema de personalização de e-mail usado no teste foi manual ou automatizado? A infraestrutura de tecnologia aguenta enviar milhões desses e-mails personalizados por dia?
* **Por quê:** Muitas vezes, um experimento funciona "na mão", mas é tecnicamente inviável ou caro demais para ser aplicado em toda a base de usuários.